### DISCUSSION TITLE
Cython 3 and pyodide-build

### DISCUSSION THREAD

**Discussion by gabrielfougeron** [pyodide/pyodide#4527] [General]
Hi,

I see in https://github.com/pyodide/pyodide/blob/main/pyodide-build/pyproject.toml that the cython dependency is currently marked as "cython<3.0".

I assume there are a lot of good reasons for that, and I'm not trying to challenge this decision.

In my own projects though, I've recently found myself to use cython 3.xx features more and more often like support for f-strings. Nothing major, but still interesting quality of life improvements.
Is a shift to / support of cython 3 planned in the future? if yes, what kind of timeframe could it be reasonably expected?

Thanks to the pyodide team for all the hard work

**Comment by ryanking13** [comment]
Oh, this is a very good catch. Thanks @gabrielfougeron.

I think cython is not required anymore as we use isolated environment now when building packages and it is package maintainers' job to declare cython as a build dependency. My guess is that we still have the cython dependency that we added before we started using the isolated build env.

A minor issue is that we use cython in some package recipes so these need to be updated

- https://github.com/pyodide/pyodide/blob/67906c3c8404e55f57a25dd06acd6ddeb9c5e6c5/packages/gensim/meta.yaml#L22
- https://github.com/pyodide/pyodide/blob/67906c3c8404e55f57a25dd06acd6ddeb9c5e6c5/packages/logbook/meta.yaml#L18

**Reply by gabrielfougeron** [reply]
> I think cython is not required anymore as we use isolated environment now when building packages and it is package maintainers' job to declare cython as a build dependency.

Got it, awesome news. Thanks a lot for the clarifications.

**Reply by hoodmane** [reply]
> My guess is that we still have the cython dependency that we added before we started using the isolated build env.

It isn't from before we used the isolated build env, the reason for Cython is because packages have a tendency to distribute their sdists with pre-cythonized code and no build-time dependency on scipy. This causes trouble when they are dependent on scipy because our scipy is not abi-compatible with ordinary scipy, so we need to recythonize. Then when Cython 3 came out a bunch of these packages broke so we capped Cython < 3. I think @mkoeppe fixed it a few days ago in #4426, and it was just an oversight that he didn't drop the cap in pyproject.toml as well.

**Reply by mkoeppe** [reply]
Yes, #4426 was a minimal fix to get rid of the global constraint Cython < 3, which affected the isolated build envs of all packages. Changing the runtime env of pyodide-build was not necessary for this.
Removing the runtime dependency of pyodie-build as done in #4528 LGTM.

**Reply by mkoeppe** [reply]
> packages have a tendency to distribute their sdists with pre-cythonized code

Fortunately this ancient practice has fallen from favor in the past years, and I'd strongly suggest to submit PRs to remove it to any projects that may still be using it

Edit: Ref to Cython upstream discussion: https://github.com/cython/cython/issues/5089

**Reply by hoodmane** [reply]
> Fortunately this ancient practice has fallen from favor in the past years, and I'd strongly suggest to submit PRs to remove it to any projects that may still be using it

Good to know! It certainly causes us headaches. @ryanking13 let's patch gensim to fix this instead of vendoring the recythonized .c file in #4528.

**Reply by ryanking13** [reply]
I see. I'll update #4528 in that way. Thanks for the information!
